Fix "Group by" error (postgreSQL)

twister65 opened the pull request:
Pull Request for Issue #17582  .

### Summary of Changes
Add m.language and l.image in the "GROUP BY" clause :
`sql_group="a.id, m.language, l.image"`

### Testing Instructions
Login as administrator.
Click on Menus


### Expected result
All menus are displayed as expected.


### Actual result
All menus are displayed as expected.


### Documentation Changes Required

twister65 committed:
Fix problems with postgresql update

twister65 committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

twister65 committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

twister65 committed:
Merge pull request #1 from joomla/staging

Update from joomla/joomla-cms:staging

twister65 committed:
Fix GROUP BY error

See Issue https://github.com/joomla/joomla-cms/issues/17582

infograf768 commented:
@izharaazmi @alikon 
Please test

izharaazmi commented on administrator/components/com_menus/presets/joomla.xml:
```diff
@@ -209,7 +209,7 @@
 			sql_where="a.client_id = 0"
 			sql_leftjoin="#__menu AS m ON m.menutype = a.menutype AND m.home = 1 LEFT JOIN #__languages AS l ON l.lang_code = m.language"
-			sql_group="a.id"
+			sql_group="a.id, m.language, l.image"
```
I think this will cause showing menu repeated for each language home. However we know this is not possible in a sane system. But we are anyway handling this case like `where count(home) > 1`. @infograf768 please see.

alikon replied:
@izharaazmi  can you explain  in plain english  what that query is supposed to fetch ?
maybe we can find an alternative way to write that query.....

izharaazmi replied:
Sure.

Basically we want to fetch a list of menus (records from "#__menu_types").
If a menu contains a default home page link (menu item with home=1)  we display an icon after the title.
If it is a home for all languages we show home icon (css clas icon-home). If for a specific language we show the relevant language flag (l.image).
Now the final tricky part - if the menu has multiple home items we show a globe icon (image:mod_languages/icon-16-language.png)

SamuelSchepp commented:
I have tested this item :white_check_mark: successfully on 2a2fbd00ae1f6c3a231a09e253d1168a5140f9a8<br /><br />Teststack:
MAMP 4.2 on macOS 10.1.2
PostgreSQL 9.6.4 on Docker 

Tested:
- Log in as administrator
- Menus -> Manage (Without patch, this action is not possible due to sql error)

Without patch:
Error page with title "Error: 0 ERROR: column "m.language" must appear in the GROUP BY clause or be used in an aggregate function LINE 1: ...E(SUM(m.home), 0) WHEN 0 THEN '' WHEN 1 THEN CASE m.language... ^"

With patch:
All menus are displayed as expected.

Tested @iCampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17620">issues.joomla.org/tracker/joomla-cms/17620</a>.</sub>

izharaazmi commented:
Please also test the following condition:

Go to #__menu table and set two or more record in a menutype to home=1 manually. Then check if the globe icon is displayed in the admin menu for that menu type.

twister65 committed:
Merge pull request #2 from joomla/staging

Update Joomla-CMS-staging

Kev1n337 commented:
@izharaazmi Could you specify further what I need to do, to test your scenario? I manually set some lines to home=1, but I don't know where and when I should see that globe icon.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17620">issues.joomla.org/tracker/joomla-cms/17620</a>.</sub>

twister65 commented:
I've set two records to home=1 in the #__menu table, and the globe icon is well displayed.
Just log in to administrator, and click on "Menus".

Kev1n337 commented:
I have tested this item :white_check_mark: successfully on f3dc52a125b8aecf79687d68670ed12b75a9032b<br /><br />When testing Joomla 3.7.5 with PostgreSQL, Administrator page displays noting but the error description: "Error page with title "Error: 0 ERROR: column "m.language" must appear in the GROUP BY clause or be used in an aggregate function LINE 1"
After applying the patch it worked as expected and all menus were displayed correctly.
@icampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17620">issues.joomla.org/tracker/joomla-cms/17620</a>.</sub>

ghost commented:
@SamuelSchepp can you please retest?

SamuelSchepp commented:
I have tested this item :white_check_mark: successfully on f3dc52a125b8aecf79687d68670ed12b75a9032b<br /><br />Still works with patch as described earlier. @icampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17620">issues.joomla.org/tracker/joomla-cms/17620</a>.</sub>

SamuelSchepp commented:
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17620">issues.joomla.org/tracker/joomla-cms/17620</a>.</sub>

ghost commented:
RTC after two successful tests.

SamuelSchepp commented:
I have tested this item :white_check_mark: successfully on f3dc52a125b8aecf79687d68670ed12b75a9032b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17620">issues.joomla.org/tracker/joomla-cms/17620</a>.</sub>

twister65 committed:
Responsive random image module

Width : in percentage.
Height : auto

twister65 committed:
Responsive Random Image Module

Width (%) : in percentage .
Height (auto) : automatic.

Quy commented on modules/mod_random_image/tmpl/default.php:
```diff
@@ -13,7 +13,7 @@
 <a href="<?php echo $link; ?>">
 <?php endif; ?>
-	<?php echo JHtml::_('image', $image->folder . '/' . $image->name, $image->name, array('width' => $image->width, 'height' => $image->height)); ?>
+	<?php echo '<img src="'.$image->folder.'/'.$image->name.'" style="max-width:'.($image->width>100?100:$image->width).'%; height:auto;" />'; ?>
```
Please follow coding style on concatenation spacing:

https://developer.joomla.org/coding-standards/php-code.html#concatenation-spacing

twister65 committed:
Responsive Random Image Module

Concatenation spacing (Joomla! coding rules)

https://forum.joomla.org/viewtopic.php?t=840822

wojsmol commented:
@twister65 Last 2 commits are in wrong branch - this should by a separate PR.

twister65 committed:
Merge branch 'staging' of https://github.com/twister65/joomla-cms into staging

twister65 committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

twister65 committed:
Revert "Responsive Random Image Module"

This reverts commit beec9ff2c4093f37ca620fa006d1d7c468b155a3.

twister65 committed:
Revert "Responsive random image module"

This reverts commit 91185f5fd67ff8c97f3eea6708291e417f49187a.

twister65 committed:
Revert "Responsive Random Image Module"

This reverts commit d9e4b8170be23f5e93dfd7222c14f8f940638f7b.
